### ISSUE TITLE
abci: Tendermint state.AppHash does not match AppHash after replay.

### ISSUE THREAD

**Issue by tglaeser** [tendermint/tendermint#425]
**BUG REPORT**

**Tendermint version** :: `692691938ccabc48f5bc312bb37f820a66634ec3`

**ABCI app** :: self-written, not publicly available

Essentially, this is a Java app similar to the Go Counter app. In the context of the error, what it does is the following:
1. Stores the received `hash` value received in callback `requestBeginBlock`
2. Stores the received `height` value received in callback `requestEndBlock`
3.  Returns those two stored values via callback `requestInfo`

**Merkleeyes version** :: n/a

**Environment**
- **OS** :: Any
- **Install tools** :: Tendermint from source

**What happened**

I'm running `tendermint node`. After killing the process and starting it again, I'm seeing the following error.
```
$ tendermint node
NOTE[03-04|20:47:24] Loaded PrivValidator                     module=types file=C:\\home\\tglaeser/.tendermint/priv_validator.json privValidator="PrivValidator{642FE53D7B76C5E81F5CEDDA3E8736FFB39B97F3 LH:0, LR:0, LS:0}"
NOTE[03-04|20:47:24] ABCI Handshake                           module=consensus appHeight=0 appHash=
NOTE[03-04|20:47:24] ABCI Replay Blocks                       module=consensus appHeight=0 storeHeight=0 stateHeight=0
NOTE[03-04|20:47:24] ConsensusReactor                         module=consensus fastSync=false
NOTE[03-04|20:47:24] Catchup by replaying consensus messages  module=consensus height=1
NOTE[03-04|20:47:24] Replay: Done                             module=consensus
NOTE[03-04|20:47:24] Starting RPC HTTP server on tcp socket 0.0.0.0:46657 module=rpcserver
NOTE[03-04|20:47:24] Started node                             module=main nodeInfo="&{PubKey:PubKeyEd25519{E6D544AF9ADF09C0B989FD828AF1EB309E885A2E416327D5C633820F1DD169B5} Moniker:anonymous Network:test-chain-9Rawjq RemoteAddr: ListenAddr:70.174.164.77:46656 Version:0.8.0 Other:[wire_version=0.6.0 p2p_version=0.3.5 consensus_version=v1/0.2.2 rpc_version=0.6.0/3 rpc_addr=tcp://0.0.0.0:46657]}"
NOTE[03-04|20:47:25] enterNewRound(1/0). Current: 1/0/RoundStepNewHeight module=consensus
NOTE[03-04|20:47:25] enterPrecommit: +2/3 prevoted proposal block. Locking module=consensus hash=03FCA8088B6377DE1844A4ED005941B1505678F2
NOTE[03-04|20:47:25] Finalizing commit of block with 0 txs    module=consensus height=1 hash=03FCA8088B6377DE1844A4ED005941B1505678F2 root=
NOTE[03-04|20:47:26] enterNewRound(2/0). Current: 2/0/RoundStepNewHeight module=consensus
NOTE[03-04|20:47:26] enterPrecommit: +2/3 prevoted proposal block. Locking module=consensus hash=CC3786BB7666B5BB5D5B47EE9BC35DC466FB8518
NOTE[03-04|20:47:26] Finalizing commit of block with 0 txs    module=consensus height=2 hash=CC3786BB7666B5BB5D5B47EE9BC35DC466FB8518 root=
NOTE[03-04|20:47:27] enterNewRound(3/0). Current: 3/0/RoundStepNewHeight module=consensus
NOTE[03-04|20:47:27] enterPrecommit: +2/3 prevoted proposal block. Locking module=consensus hash=BBC2AFD674CE6CB652E4F4BE9C6C3DCA1BC15524
NOTE[03-04|20:47:28] Finalizing commit of block with 0 txs    module=consensus height=3 hash=BBC2AFD674CE6CB652E4F4BE9C6C3DCA1BC15524 root=
NOTE[03-04|20:47:29] enterNewRound(4/0). Current: 4/0/RoundStepNewHeight module=consensus
NOTE[03-04|20:47:29] enterPrecommit: +2/3 prevoted proposal block. Locking module=consensus hash=07D539BBC5E0E62DC951D810B5D826AD204AE902
NOTE[03-04|20:47:29] Finalizing commit of block with 0 txs    module=consensus height=4 hash=07D539BBC5E0E62DC951D810B5D826AD204AE902 root=
captured interrupt, exiting...
NOTE[03-04|20:47:29] Stopping Node                            module=node

$ tendermint node
NOTE[03-04|20:47:30] Loaded PrivValidator                     module=types file=C:\\home\\tglaeser/.tendermint/priv_validator.json privValidator="PrivValidator{642FE53D7B76C5E81F5CEDDA3E8736FFB39B97F3 LH:4, LR:0, LS:3}"
NOTE[03-04|20:47:30] ABCI Handshake                           module=consensus appHeight=4 appHash=07D539BBC5E0E62DC951D810B5D826AD204AE902
NOTE[03-04|20:47:30] ABCI Replay Blocks                       module=consensus appHeight=4 storeHeight=4 stateHeight=4
panic: Tendermint state.AppHash does not match AppHash after replay. Got 07D539BBC5E0E62DC951D810B5D826AD204AE902, expected

goroutine 1 [running]:
github.com/tendermint/tendermint/consensus.(*Handshaker).checkAppHash(0xc0428b47b0, 0xc042456620, 0x14, 0x14, 0x6, 0x6)
        C:/home/projects/go/src/github.com/tendermint/tendermint/consensus/replay.go:340 +0x26d
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(0xc0428b47b0, 0xc042456620, 0x14, 0x14, 0x4, 0xf18860, 0xc042464360, 0x14, 0x14, 0x0, ...)
        C:/home/projects/go/src/github.com/tendermint/tendermint/consensus/replay.go:255 +0x408
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc0428b47b0, 0xf18860, 0xc042464360, 0x0, 0x0)
        C:/home/projects/go/src/github.com/tendermint/tendermint/consensus/replay.go:209 +0x458
github.com/tendermint/tendermint/proxy.(*multiAppConn).OnStart(0xc042464360, 0xc042127880, 0x15)
        C:/home/projects/go/src/github.com/tendermint/tendermint/proxy/multi_app_conn.go:100 +0x219
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common.(*BaseService).Start(0xc042464360, 0x0, 0x0, 0x0)
        C:/home/projects/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common/service.go:96 +0x2c9
github.com/tendermint/tendermint/node.NewNode(0xf195a0, 0xc04246fe80, 0xc042133860, 0xf0b7c0, 0xc042461b30, 0x0)
        C:/home/projects/go/src/github.com/tendermint/tendermint/node/node.go:73 +0x599
github.com/tendermint/tendermint/node.NewNodeDefault(0xf195a0, 0xc04246fe80, 0x1)
        C:/home/projects/go/src/github.com/tendermint/tendermint/node/node.go:53 +0xc1
main.run_node(0xf195a0, 0xc04246fe80)
        C:/home/projects/go/src/github.com/tendermint/tendermint/cmd/tendermint/run_node.go:49 +0x58d
main.main()
        C:/home/projects/go/src/github.com/tendermint/tendermint/cmd/tendermint/main.go:46 +0x2c0
```

**What you expected to happen**:

Well, it seems that `appHeight == storeHeight == stateHeight` and `appHash == hash` for the last finalized block, so it is not clear why the Tendermint node cannot resume.

**How to reproduce it**:

Just testing the same app after #409 was fixed.

**Comment by ebuchman** [comment]
`panic: Tendermint state.AppHash does not match AppHash after replay. Got 07D539BBC5E0E62DC951D810B5D826AD204AE902, expected` 

Note Tendermint is expecting an empty app hash here ... do you return the app hash on `Commit` ? Tendermint expects the app hash returned by requestInfo to be the same as the app hash returned by the last call to requestCommit before the crash. If you're not returning anything on commit, then that would explain why Tendermint is unhappy, and thus it wouldn't be a bug - just perhaps horrible documentation, which is maybe almost as bad as a bug :P . Otherwise, this is more serious.

**Comment by tglaeser** [comment]
Yes, I did notice that Tendermint did expect an empty hash here. And no, I did not return the hash on commit, instead I did return the `tx` value as I learned from the Go Counter app.

In this context, what is the meaning of value `root` from the Tendermint console logging?
```
NOTE[03-05|00:30:22] Finalizing commit of block with 0 txs    module=consensus height=160 hash=2ACFBC7832CE533A2D618C2ADE3003B376E5A916 root=E5601A63217D250B4E02562AAA3EB2E099961A34
```

**Comment by tglaeser** [comment]
OK, I found this in the documentation ... _The Commit response includes a byte array, which is the deterministic state root of the application. It is included in the header of the next block. It can be used to provide easily verified Merkle-proofs of the state of the application._

If I understand this correctly than the value of `root` for block `n+1` is equal to the value of `hash` for block `n`.

Thanks for your help. Not a bug, user error. Sorry for bothering you...

**Comment by ebuchman** [comment]
Ah yes - the Counter is a bad example for this since it doesn't return a real Merkle root, and doesn't do anything for requestInfo. If you return an empty requestInfo, Tendermint will replay *all* blocks on restart.

We typically expect apps to return a state root on Commit, which is the Merkle root of the application state. This value is the `AppHash` in the next block, sometimes just called `root` (as you noticed).  So indeed, in `Finalizing commit of block with 0 txs    module=consensus height=160 hash=2ACFBC7832CE533A2D618C2ADE3003B376E5A916 root=E5601A63217D250B4E02562AAA3EB2E099961A34`, the `hash` is the hash of the block and the `root` is what was returned from `Commit` after processing the *previous* block.

It's a bit confusing. We should probably be more careful in our use of the terms `hash` and `root`, and talk explicitly about `block_hash` (the hash passed in BeginBlock) and `last_app_hash` (for block N, this is returned after calling Commit for block N-1) and `app_hash` (the result of Commit for the current block). In your case, it's a bit more confusing because you're using the block hash (passed in BeginBlock) as the app hash (by returning it on requestInfo and maybe now on requestCommit), which is not really correct.

In conclusion, the thing hash returned for requestInfo should be the same as the hash returned in the last call to requestCommit.

> Thanks for your help. Not a bug, user error. Sorry for bothering you...

Cool, not a problem. I'll close the issue and we'll try and improve the docs/naming to make this clearer moving forward.

**Comment by tglaeser** [comment]
I'm sorry, but I need to follow-up on this. Naming is important ...

I found the following definition in `types.proto`
```
message ResponseInfo {
	string data = 1;
	string version = 2;
	uint64 last_block_height = 3;
	bytes last_block_app_hash = 4;
}
```
What does `last_block_app_hash` mean? Is this the `block_hash` or the `last_app_hash` or potentially both?

**Comment by ebuchman** [comment]
Oh boy - it's amazing what one takes for granted when knee deep in the trenches. 

It's the `app_hash` returned by Commit after running the `last_block`. 

In other words, `last_block_height` is the height of the last block that the app successful ran `Commit` for, and `last_block_app_hash` is the corresponding return value of that `Commit`.

**Comment by tglaeser** [comment]
I'm voting to rename `last_block_app_hash` to `last_app_hash`. This would match what you return when calling `curl http://localhost:46657/status` ... almost ... `last_*` vs. `latest_*`

**Comment by ebuchman** [comment]
Except it's actually distinct. If we've committed block N, The `latest_app_hash` in /status is the app hash included in block N, which is the result of Commit on block N-1. The `last_block_app_hash` is the result of Commit on block N. 

Still, maybe it makes sense to call it `last_app_hash`

**Comment by laitkor** [comment]
I am getting the same error when running from root user, please suggest a possible solution for the same.Thankyou

**Comment by EnchanterIO** [comment]
same error here with version 0.14

**Comment by vanipolnedi** [comment]
[root@localhost ~]# tendermint node   
I[10-24|07:00:26.787] Starting multiAppConn                        module=proxy impl=multiAppConn
I[10-24|07:00:26.787] Starting socketClient                        module=abci-client connection=query impl=socketClient
I[10-24|07:00:26.787] Starting socketClient                        module=abci-client connection=mempool impl=socketClient
I[10-24|07:00:26.787] Starting socketClient                        module=abci-client connection=consensus impl=socketClient
I[10-24|07:00:26.790] ABCI Handshake                               module=consensus appHeight=13731 appHash=31343330383037653361363239303733666432376263393630653438356239383561306235376365306461343531383036663662383332336534643732376534
I[10-24|07:00:26.790] ABCI Replay Blocks                           module=consensus appHeight=13731 storeHeight=0 stateHeight=0
panic: Tendermint state.AppHash does not match AppHash after replay. Got 31343330383037653361363239303733666432376263393630653438356239383561306235376365306461343531383036663662383332336534643732376534, expected 

goroutine 1 [running]:
github.com/tendermint/tendermint/consensus.checkAppHash(0xc42073c4c0, 0x11, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/consensus/replay.go:426 +0x1c7
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(0xc42079e000, 0xc42073c4c0, 0x11, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/consensus/replay.go:301 +0xd79
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc42079e000, 0xf94c40, 0xc420118900, 0x4, 0xf8c460)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/consensus/replay.go:246 +0x3e5
github.com/tendermint/tendermint/proxy.(*multiAppConn).OnStart(0xc420118900, 0xc42073c6c0, 0x15)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/proxy/multi_app_conn.go:108 +0x527
github.com/tendermint/tendermint/libs/common.(*BaseService).Start(0xc420118900, 0x0, 0x0)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/libs/common/service.go:130 +0x3bd
github.com/tendermint/tendermint/node.NewNode(0xc4201a3b00, 0xf8d9e0, 0xc42014ac60, 0xf80da0, 0xc420799b90, 0xc4207c30e0, 0xf15c88, 0xf15c90, 0xf8c460, 0xc4207c43c0, ...)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/node/node.go:187 +0x693
github.com/tendermint/tendermint/node.DefaultNewNode(0xc4201a3b00, 0xf8c460, 0xc4207c43c0, 0x0, 0xc42009db20, 0x5887f0)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/node/node.go:83 +0x1cf
github.com/tendermint/tendermint/cmd/tendermint/commands.NewRunNodeCmd.func1(0xc4200b8d80, 0x1492148, 0x0, 0x0, 0x0, 0x0)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/cmd/tendermint/commands/run_node.go:53 +0x5e
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200b8d80, 0x1492148, 0x0, 0x0, 0xc4200b8d80, 0x1492148)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:698 +0x46d
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1468800, 0xcc9de0, 0x12, 0x20)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:783 +0x2e4
github.com/tendermint/tendermint/vendor/github.com/spf13/cobra.(*Command).Execute(0x1468800, 0x20, 0xc42005de98)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/vendor/github.com/spf13/cobra/command.go:736 +0x2b
github.com/tendermint/tendermint/libs/cli.Executor.Execute(0x1468800, 0xf17be0, 0x2, 0xc420019ea0)
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/libs/cli/setup.go:89 +0x4e
main.main()
        /home/jslave/workspace/02.Fetch.Binary@2/go/src/github.com/tendermint/tendermint/cmd/tendermint/main.go:45 +0x24d



any suggestions please

**Comment by ebuchman** [comment]
@vanipolnedi what version are you using?

**Comment by vanipolnedi** [comment]
@ebuchman tendermint version : 0.22.8-40d6dc2e
bigchaindb version : bigchaindb 2.0.0b7

**Comment by vanipolnedi** [comment]
@ebuchman  one more doubt i have ....can we able to retrieve multiple values stored under same key or only updated value of that particular key in tendermint ??? ...if we cant retrieve multiple values of same key ...is there any alternative to this ??

thanks in advance

**Comment by vanipolnedi** [comment]
@ebuchman  I[10-24|07:00:26.790] ABCI Replay Blocks module=consensus appHeight=13731 storeHeight=0 stateHeight=0    ............... then, what is the solution to reset appHeight ...any suggestions please

**Comment by ebuchman** [comment]
That depends on your app. The app tracks its own height and returns it in ResponseInfo.LastBlockHeight. What app are you using?

**Comment by vanipolnedi** [comment]
@ebuchman  i'm using kvstore application

**Comment by ebuchman** [comment]
Did you reset the Tendermint data without resetting the BigChainDB data? Looks like BigChainDB is working from old data from and old blockchain while Tendermint has started from a fresh slate ..

**Comment by vanipolnedi** [comment]
@ebuchman after resetting BigChainDB data , it's working fine .thank you :)

**Comment by KrishnaPG** [comment]
Came across this while searching for answer for the error: `Tendermint state.AppHash does not match AppHash after replay. Got , expected 0800000000000000`

I am using the example code from [acbi counter](https://github.com/tendermint/js-abci/blob/master/example/counter.js ) which does not implement the `commit` method, which apparently seems to be required for this to work. Besides, the `info` method has very simple implementation that does not return a value for the `lastBlockAppHash`
````
  info (request) {
    return {
      data: 'Node.js counter app',
      version: '0.0.0',
      lastBlockHeight: 0,
      lastBlockAppHash: Buffer.alloc(0)
    }
  }
`````
The question is:

- how to return valid values for the `lastBlockHeight` and `lastBlockAppHash`, since tendermint node may not be running at the time the acbi app is getting started
- would returning valid values from the `reuqest info` method avoid the app hash not matching problem (without having to know and return the last valid app root hash in the `commit` method?)

In other words, how can the acbi app know what was the last app root hash and return it properly (since the tendermint node is just getting started and has not completed the replay?)

**Comment by ebuchman** [comment]
Can you give us more info on the sequence of events that led to this error? Ideally a description of how to replicate ...

**Comment by KrishnaPG** [comment]
Thank you @ebuchman 

In my case, I was trying to understand the whole ABCI and tendermint concepts. More importantly, trying to understand how to validate and store the block data in custom backend store / DB.

Following the instructions, started the tendermint with the command `tendermint node --proxy_app=kvstore` after installation as suggested in the [docs](https://tendermint.com/docs/introduction/quick-start.html#initialization). Entered few data blocks using the CURL as per the docs.

Then tried to use the `counter.js` example as the ABCI app. That is when the error started. Now I believe this happend because the ABCI app changed (from kvstore to counter.js) - which is fine. 

What I do not understand is, how can the `counter.js` pickup from where the earlier ABCI app left and continue from there? In this case, since the apps are different, it may not be possible. But in general, how can `counter.js` app access the last hash data from the tendermint storage so that it can resume from the previous instance? The crash-recovery doc you have suggested says:

> On startup, Tendermint calls the Info method on the Info Connection to get the latest committed state of the app. The app MUST return information consistent with the last block it succesfully completed Commit for.

Is it expected that ABCI app should maintain its own storage (database / file) and keep track of the state separately? Is there anyway the app can access the Tendermint block storage to get that last block information?

In general, how can the ABCI app take control of the Tendermint' storage (e.g. use custom database for storing the blocks)?

**Comment by ebuchman** [comment]
> What I do not understand is, how can the counter.js pickup from where the earlier ABCI app left and continue from there? In this case, since the apps are different, it may not be possible.

Ah, yes, this isn't really supported as it doesn't really make sense - the state machine shouldn't just suddenly change in the midst of its life.

> Is it expected that ABCI app should maintain its own storage (database / file) and keep track of the state separately?

Yes exactly. If it doesn't, then it will tell Tendermint that the latest height it knows about is 0, in which case Tendermint will replay all the blocks it has from 0, letting the app recompute everything. This is perfectly acceptable if you can tolerate the startup latency from replaying all the block :) 

> In general, how can the ABCI app take control of the Tendermint' storage (e.g. use custom database for storing the blocks)?

There are effectively two DB systems - one used by the app to store state, and one used by Tendermint to store blocks. For the app, you can use whatever you want. For Tendermint, there are a few built in options, but to add more you'd have to create your own import the tendermint `node` package to initialize a node using the DB of your choosing. 

Hope that helps!

**Comment by KrishnaPG** [comment]
Thank you @ebuchman I believe the approach of ABCI app returning 0, and Tendermint replaying all the blocks makes perfect sense.

Will try to look into more on how to make the Tendermint use some other databases to store the blockchain blocks. 

If one were to, say, index some meta data about each block into, say ElasticSearch or some other DB, as and when a new block is added to the chain in realtime, what would be the best approach? The ABCI app, or Tendermint node package?

**Comment by ebuchman** [comment]
We typically recommend using the Tendermint node's RPC interface. See the rpc docs at https://tendermint.com/docs/tendermint-core/rpc.html and the websockets docs at https://tendermint.com/docs/app-dev/subscribing-to-events-via-websocket.html.

You can also see how Bez is doing it for the Cosmos Hub (piping data into Postgres): https://github.com/alexanderbez/juno

Shall we close this again?

**Comment by KrishnaPG** [comment]
Yes, please - can be closed. Thank you for the pointers.
